# Caiza Klar Laboratorio 2

## Limpieza de datos 08/07/2022

In [2]:
'''Importamos las respectivas librerias para el desarrollo del laboratorio'''
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns
from sklearn.impute import SimpleImputer

In [5]:
'''Cargamos el conjunto de datos contenidos dentro del archivo csv'''
df = pd.read_csv('data_act_01.csv', engine = 'python',sep=";")

In [7]:
'''Imprimiremos las primeras 5 filas del archivo'''
df.head()

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
0  160903280     Assault / Battery  2016-03-30T00:00:00    18:42   
1  160912272    Homeless Complaint  2016-03-31T00:00:00    15:31   
2  160912590             Susp Info  2016-03-31T00:00:00    16:49   
3  160912801                Report  2016-03-31T00:00:00    17:38   
4  160912811                   594  2016-03-31T00:00:00    17:42   

          CallDateTime Disposition                  Address           City  \
0  2016-03-30T18:42:00         REP  100 Block Of Chilton Av  San Francisco   
1  2016-03-31T15:31:00         GOA  2300 Block Of Market St  San Francisco   
2  2016-03-31T16:49:00         GOA  2300 Block Of Market St  San Francisco   
3  2016-03-31T17:38:00         GOA      500 Block Of 7th St  San Francisco   
4  2016-03-31T17:42:00         REP       Beale St/bryant St  San Francisco   

  State AgencyId  Range      AddressType  
0    CA        1    NaN  Premise Address  
1    CA        1    NaN  Premise Address  
2    CA        1    NaN  Premise Address  
3    CA        1    NaN  Premise Address  
4    CA        1    NaN     Intersection

In [8]:
'''Verificamos si las variables fueron cargadas correctamente a traves
de la funcion info y ded igual manera verificamos el tipo de variables'''
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10051 entries, 0 to 10050
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   CrimeId                10051 non-null  int64  
 1   OriginalCrimeTypeName  10051 non-null  object 
 2   OffenseDate            10051 non-null  object 
 3   CallTime               10051 non-null  object 
 4   CallDateTime           10051 non-null  object 
 5   Disposition            10051 non-null  object 
 6   Address                10051 non-null  object 
 7   City                   9730 non-null   object 
 8   State                  10048 non-null  object 
 9   AgencyId               10051 non-null  object 
 10  Range                  0 non-null      float64
 11  AddressType            10051 non-null  object 
dtypes: float64(1), int64(1), object(10)
memory usage: 942.4+ KB


In [9]:
'''Imprimimos la cantidad de datos que posee el dataset en
un funcion del número de filas y columnas'''
df.shape

(10051, 12)

In [10]:
'''Verificamos si el dataset posee valores nulos mediante la suma'''
df.isnull().sum()

CrimeId                      0
OriginalCrimeTypeName        0
OffenseDate                  0
CallTime                     0
CallDateTime                 0
Disposition                  0
Address                      0
City                       321
State                        3
AgencyId                     0
Range                    10051
AddressType                  0
dtype: int64

# Error 1
# Eliminacion columnas "Range" y "DateTime"

In [13]:
'''Eliminacion columna range'''
df= df.drop(columns=['Range'], errors='ignore')
df

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
0      160903280     Assault / Battery  2016-03-30T00:00:00    18:42   
1      160912272    Homeless Complaint  2016-03-31T00:00:00    15:31   
2      160912590             Susp Info  2016-03-31T00:00:00    16:49   
3      160912801                Report  2016-03-31T00:00:00    17:38   
4      160912811                   594  2016-03-31T00:00:00    17:42   
...          ...                   ...                  ...      ...   
10046  160964210      Well Being Check  2016-04-05T00:00:00    23:38   
10047  160964216          Traffic Stop  2016-04-05T00:00:00    23:43   
10048  160964227          Traffic Stop  2016-04-05T00:00:00    23:48   
10049  160964229          Passing Call  2016-04-05T00:00:00    23:48   
10050  160964249          Traffic Stop  2016-04-05T00:00:00    23:54   

              CallDateTime   Disposition                  Address  \
0      2016-03-30T18:42:00           REP  100 Block Of Chilton Av   
1      2016-03-31T15:31:00           GOA  2300 Block Of Market St   
2      2016-03-31T16:49:00           GOA  2300 Block Of Market St   
3      2016-03-31T17:38:00           GOA      500 Block Of 7th St   
4      2016-03-31T17:42:00           REP       Beale St/bryant St   
...                    ...           ...                      ...   
10046  2016-04-05T23:38:00           GOA    Hollister Av/hawes St   
10047  2016-04-05T23:43:00           ADV   California St/mason St   
10048  2016-04-05T23:48:00           CIT   Rosella Ct/onondaga Av   
10049  2016-04-05T23:48:00  Not recorded          0 Blk Herbst Rd   
10050  2016-04-05T23:54:00           ADV        Grant Av/geary St   

                City State AgencyId      AddressType  
0      San Francisco    CA        1  Premise Address  
1      San Francisco    CA        1  Premise Address  
2      San Francisco    CA        1  Premise Address  
3      San Francisco    CA        1  Premise Address  
4      San Francisco    CA        1     Intersection  
...              ...   ...      ...              ...  
10046  San Francisco    CA        1     Intersection  
10047  San Francisco    CA        1     Intersection  
10048  San Francisco    CA        1     Intersection  
10049  San Francisco    CA        1     Geo-Override  
10050  San Francisco    CA        1     Intersection  

[10051 rows x 11 columns]

In [26]:
'''Eliminacion columa dateTime'''
df= df.drop(columns=['CallDateTime'], errors='ignore')
'''Imprimimos el nuevo dataset'''
df.head()

CrimeId OriginalCrimeTypeName          OffenseDate CallTime Disposition  \
0  160903280     Assault / Battery  2016-03-30T00:00:00    18:42         REP   
1  160912272    Homeless Complaint  2016-03-31T00:00:00    15:31         GOA   
2  160912590             Susp Info  2016-03-31T00:00:00    16:49         GOA   
3  160912801                Report  2016-03-31T00:00:00    17:38         GOA   
4  160912811                   594  2016-03-31T00:00:00    17:42         REP   

                   Address           City State AgencyId      AddressType  
0  100 Block Of Chilton Av  San Francisco    CA        1  Premise Address  
1  2300 Block Of Market St  San Francisco    CA        1  Premise Address  
2  2300 Block Of Market St  San Francisco    CA        1  Premise Address  
3      500 Block Of 7th St  San Francisco    CA        1  Premise Address  
4       Beale St/bryant St  San Francisco    CA        1     Intersection

## Verificacion error 1

Una vez realizada la importacion de todas las librerias, ademas de cargar y analizar el dataset, logramos identificar el primer error correspondiente a diversas columnas que no son relevantes. Como primer error en el dataset se identificaron varias columnas que poseen datos vacios o nulos siendo esta Range y una columna que es redundante para el dataset siendo CallDateTime, ya que existen columnas encargadas de ofrecer informacion de la hora y fecha. Ambas columnas ocupan espacio dentro del dataset y pueden desembocar en errores futuros al manejar dicho dataset. Para el manejo de columnas innecesarias, se hizo uso de la funcion drop el cual es el encargado de elimnar las columnas que se agreguen. Finalmente logramos imprimir el nuevo dataset con las columnas ya eliminadas a traves de su impresion mediande sample, el cual permite imprimir cierta cantidad de columnas aleatorias.

In [27]:
'''Nuevo df'''
df.sample(3)

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
3120  160931436                22500e  2016-04-02T00:00:00    11:42   
5755  160942084     Suspicious Person  2016-04-03T00:00:00    14:22   
2481  160930261        Noise Nuisance  2016-04-02T00:00:00     1:44   

     Disposition                  Address           City State AgencyId  \
3120         GOA  1100 Block Of Moraga St  San Francisco    CA        1   
5755         REP     Haight St/shrader St  San Francisco    CA        1   
2481         GOA    Ofarrell St/larkin St  San Francisco    CA        1   

          AddressType  
3120  Premise Address  
5755     Intersection  
2481     Intersection

# Error 2
# Rellenar columna "city" debido valores nulos

In [28]:
'''Rellenamos la columna city por valores N/A ya que al poseer valores nulos pueden afectar al dataset'''
df[['City', 'State']] = df[['City','State']].fillna(value="N/A")
'''Imprimimos el nuevo dataset'''
df.head()

CrimeId OriginalCrimeTypeName          OffenseDate CallTime Disposition  \
0  160903280     Assault / Battery  2016-03-30T00:00:00    18:42         REP   
1  160912272    Homeless Complaint  2016-03-31T00:00:00    15:31         GOA   
2  160912590             Susp Info  2016-03-31T00:00:00    16:49         GOA   
3  160912801                Report  2016-03-31T00:00:00    17:38         GOA   
4  160912811                   594  2016-03-31T00:00:00    17:42         REP   

                   Address           City State AgencyId      AddressType  
0  100 Block Of Chilton Av  San Francisco    CA        1  Premise Address  
1  2300 Block Of Market St  San Francisco    CA        1  Premise Address  
2  2300 Block Of Market St  San Francisco    CA        1  Premise Address  
3      500 Block Of 7th St  San Francisco    CA        1  Premise Address  
4       Beale St/bryant St  San Francisco    CA        1     Intersection

## Verificacion Error 2

Posteriormente se logro identificar un nuevo error dentro del dataset pertenciente a la columna city. Como se pudo analizar en pasos anteriores, al momento de realizar la suma de valores nulos dentro del dataset aparecio la columna city y range. Como se pudo observar la columna city, poseia un total  de 321 datos nulos y esto se modificara gracias a la funcion fillna. Como se puede apreciar en el paso anterior se modifico los valores nulos para que imprima N/A ya que al ser valores nulos, no se puede comprender la ciudad de donde proviene dicha informacion, por eso no se rellenan de manera hipotetica con cualquier otra ciudad. A continuacion, se puede apreciar como veriricamos el dataset nuevamente a traves de una suma de valores nulos y ya no existen valores como tal. Ademas a traves de la funcion groupby y agregamos la columna de ciudad podemos imprimir cada una de las ciudad correspondientes y existe una nueva que es N/A la cual fue agregada por nosotros.

In [29]:
'''Verificamos nuevamente el dataset'''
df.isnull().sum()

CrimeId                  0
OriginalCrimeTypeName    0
OffenseDate              0
CallTime                 0
Disposition              0
Address                  0
City                     0
State                    0
AgencyId                 0
AddressType              0
dtype: int64

In [30]:
'''Verificamos las ciudades dentro de la columna city y el nuevo datos N/A'''
df.groupby('City').size()

City
 S                  1
Brisbane            1
Daly City           5
N/A               321
Presidio            3
SAN FRANCISCO       1
San Francisco    9665
Treasure Isla      51
Yerba Buena         3
dtype: int64

# Error 3
# Control de ciudades, columna "city"

In [32]:
'''Una vez terminado el error 2 podemos apreciar como existe una ciudad S, 
ademas de varias ciudades que San fracisco se escribio de manera repetida, 
corregiremos este error llamandolos a los 3 por unicamente San Francisco'''
df['City']= df['City'].replace(' S','San Francisco')
df['City'] = df['City'].replace('SAN FRANCISCO','San Francisco');
df.sample(3)

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
8857  160953826    Homeless Complaint  2016-04-04T00:00:00    21:42   
2870  160930981     Suspicious Person  2016-04-02T00:00:00     9:17   
364   160920459          Passing Call  2016-04-01T00:00:00     5:50   

     Disposition                Address           City State AgencyId  \
8857         ADV  Merlin St/harrison St  San Francisco    CA        1   
2870         GOA      Larkin St/bush St  San Francisco    CA        1   
364          ADV     Block Of Haight St  San Francisco    CA        1   

          AddressType  
8857     Intersection  
2870     Intersection  
364   Premise Address

## Verificacion Error 3

Anteriormente en el paso numero 2 se pudo rellenar datos vacios que poseen nuestra columna ciudad, sin embargo esta posee mas errores que no se deben pasar por alto. Si se aprecia claramente en la verificacion del error 2, la columna city, posee una ciudad llamada unicamente "S" podemos intuir que pertenece a San francisco ya que no existe una ciudad como tal llamada "S". Ademas se aprecia como la ciudad de San francisco, esta escrita dos veces de manera diferente, esto se arreglara reemplazanso aquelas ciudad de S y sana francisco en una sola. Como se puede apreciar, al momento de realziar la funcion groupby, las ciudad "S" ya no existe, ademas que San francisco es una sola y estas tres ciudades fueron unidas en una sola.

In [34]:
'''Verificamos las ciudades dentro de la columna city '''
df.groupby('City').size()

City
Brisbane            1
Daly City           5
N/A               321
Presidio            3
San Francisco    9667
Treasure Isla      51
Yerba Buena         3
dtype: int64

# Error 4
# Control de tipo de direcciones, columna "AddressType"

In [43]:
'''Verificamos los tipos de direcciones '''
df.groupby('AddressType').size()

AddressType
1                     3
Common Location     818
Geo-Override        469
Intersection       3701
Intersectioon         1
Premise Address    5059
dtype: int64

In [48]:
'''Remplazaamos la el tipo de direccion "1" por Common Location a traves de replace'''
df['AddressType']= df['AddressType'].replace('1','Common Location')
df.sample(3)

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
5563  160941747                    Dw  2016-04-03T00:00:00    12:40   
5892  160942335          Traffic Stop  2016-04-03T00:00:00    15:34   
3956  160932921              Burglary  2016-04-02T00:00:00    18:31   

     Disposition                   Address           City State AgencyId  \
5563         CAN  1300 Block Of Vallejo St  San Francisco    CA        1   
5892         CIT          6th St/jessie St  San Francisco    CA        1   
3956         UTL   1000 Block Of Market St  San Francisco    CA        1   

          AddressType  
5563  Premise Address  
5892     Intersection  
3956  Premise Address

## Verificacion Error 4

Como se puede apreciar en nuestra primera consulta de caracter groupby realizada anteriormente hacia la columa "AddressType", esta columa posee diversos tipos de direcciones. Sin embargo, se puede obsrva que existe una direccion denominada unicamente "1", este valor es consideradom erroneo ya que no sigue la logica de las demas direcciones al ser unicamente un valor numerico. Para corregir este error se todo en consideracion este tipo de direccion por el tipo "Common Location" ya que se prevee que se trato de señalar este campo que este ocupa el primero lugar de la lista. De este modo se reemplazo modifico el campo a traves de la funcion replace asignando el noombre Common Location

In [50]:
'''Verificamos los tipos de direcciones ya modificados'''
df.groupby('AddressType').size()

AddressType
Common Location     821
Geo-Override        469
Intersection       3701
Intersectioon         1
Premise Address    5059
dtype: int64